### ISSUE TITLE
PubPoly Check fails when marshalling/ unmarshalling

### ISSUE THREAD

**Issue by WEB3-GForce** [DeDiS/crypto#35]
@davidiw 
@bford 
@jackowitzd2

In my prifi commit for life insurance, I am experiencing a strange bug in regards to public polynomials. I am passing public polynomials into messages that I then marshal/ unmarshal to send across the network. When I do so, the public polynomial no longer successfully validates shares.

As quick terminology:

   msg = the message before marshalling
   newMsg = the message after marshalling

What makes this even more interesting is the following:

    - msg properly validates msg.Share as well as newMsg.Share. So, the secrets seem to be
    intact.

     - Comparing msg.PubCommit and newMsg.PubCommit, they are not only equal according
    to the Equal function but also marshal to the same byte arrays.

Nevertheless, newMsg.PubCommit fails to recognize newMsg.share.

As for other notes of information:

- I am using abstract.Write / abstract.Read to do my marshalling / unmarshalling.

Here's the code:

// These messages are used to send insurance requests. A node looking for an
// insurance policy will send these to other nodes to ask them to become
// insurers.
type RequestInsuranceMessage struct {
	// The public key of the insured.
	PubKey abstract.Point
	
	// The number of the share being sent.
	ShareNumber *nist.Int

	// The private share to give to the insurer
	Share abstract.Secret

	// The public polynomial used to verify the share
	PubCommit *poly.PubPoly
}

/* Creates a new insurance request message
 *
 * Arguments:
 *	s  = the shared secret to give to the insurer.
 *      pc = the public polynomial to check the share against.
 *
 * Returns:
 *	A new insurance request message.
 */
func (msg *RequestInsuranceMessage) createMessage(p abstract.Point, shareNum int,
	s abstract.Secret, pc *poly.PubPoly) *RequestInsuranceMessage {
	msg.PubKey = p
	msg.ShareNumber = nist.NewInt(int64(shareNum), big.NewInt(int64(math.MaxInt64)))
	msg.Share = s
	msg.PubCommit = pc
	return msg
}

// Encodes the message for sending.
func (msg *RequestInsuranceMessage) MarshalBinary() ([]byte, error) {
	b := bytes.Buffer{}
	abstract.Write(&b, msg, INSURE_GROUP)
	return b.Bytes(), nil
//	return protobuf.Encode(msg);
}

// Decodes a message received.
// NOTE: In order to be encoded properly, public polynomials need to be
// initialized with the right group and minimum number of shares.
func (msg *RequestInsuranceMessage) UnmarshalBinary(data []byte) (*RequestInsuranceMessage, error) {
	msg.PubCommit = new(poly.PubPoly)
	msg.PubCommit.Init(INSURE_GROUP, TSHARES, nil)
	msg.PubKey = KEY_SUITE.Point()
	msg.Share = INSURE_GROUP.Secret().Pick(random.Stream)
	msg.ShareNumber = nist.NewInt(int64(0), big.NewInt(int64(math.MaxInt64)))
	b := bytes.NewBuffer(data)
	err := abstract.Read(b, msg, INSURE_GROUP)
	if err != nil {
		panic(err)
	}
	return msg, err
//	return 	msg, protobuf.Decode(data, msg)

}

Here's how I create the secret and public polynomial I pass into it:

var secret    = INSURE_GROUP.Secret().Pick(random.Stream)

var pripoly   = new(poly.PriPoly).Pick(INSURE_GROUP, TSHARES, secret, random.Stream)

// Used to initialize the public commit polynomial.
func producePubPoly() *poly.PubPoly {
	testPubPoly := new(poly.PubPoly)
	testPubPoly.Init(INSURE_GROUP, n, nil)
	return testPubPoly.Commit(pripoly, nil)
}

This is the piece of code that fails:

!newMsg.PubCommit.Check(int(newMsg.ShareNumber.V.Int64()), newMsg.Share)

Whereas the following succeed

!msg.PubCommit.Check(int(msg.ShareNumber.V.Int64()), msg.Share)
!msg.PubCommit.Check(int(newMsg.ShareNumber.V.Int64()), newMsg.Share)

Does anyone happen to know if this is a bug on my end or on the crypto side?

**Comment by bford** [comment]
As you can see, cut-and-pasting substantial amounts of code into a github message field doesn't seem to produce very readable results; a better way to discuss pieces of code like this is just to check it in - into an ephemeral branch if it's not appropriate (yet) to put into any development branch - and refer to that.   That way we can do line-comments and such.

I'm kind of surprised your msg.MarshalBinary() isn't creating an infinite-stack-depth loop: it passes a pointer to msg to abstract.Write, which might notice that msg implements the BinaryMarshaler interface can call MarshalBinary() on it instead of the "fallback" method of introspection, which will re-invoke your msg.MarshalBinary method, which calls abstract.Write on it again...  I guess perhaps currently only the protobuf encoding mechanism consistently detects when objects to be written implement the BinaryMarshaler interface, but that's probably a bug in abstract.Write (and abstract.Read), which - when fixed - will make your code loop infinitely.

In general, from this struct definition it doesn't like you should need to implement MarshalBinary() on the msg type at all; just let the caller (whoever wants the struct marshaled) pass the msg-pointer directly to abstract.Write, which will use reflection to figure out how to marshal it.  You should only need to implement MarshalBinary() on types that *need* to "marshal themselves" for some reason and for which reflection-based marshaling won't work (e.g., structs whose members are private and/or special in other ways, custom encodings, etc.).

It's hard to actually help you debug the real marshaling/unmarshaling problem without being able to checkout the code myself and run it - can you commit it to a branch somewhere (perhaps ephemeral) with a test-function that produces the behavior you're seeing, so that I can run and play with it?  Thanks.

**Comment by WEB3-GForce** [comment]
The code is checked in via the insurance branch I am working on in prifi.
The lastest branch contains the code along with a test where I produce the
behavior that I mentioned.

Within policyMessage_test.go, TestRequestInsuranceMarshallUnMarshall is the
function that tests the behavior.

In the validation, the following check fails:

  !newMsg.PubCommit.Check(int(newMsg.ShareNumber.V.Int64()), newMsg.Share)

After which I do some more testing to figure out the nature of the failure.

On Thu, Feb 19, 2015 at 4:14 PM, Bryan Ford <notifications@github.com>
wrote:

> As you can see, cut-and-pasting substantial amounts of code into a github
> message field doesn't seem to produce very readable results; a better way
> to discuss pieces of code like this is just to check it in - into an
> ephemeral branch if it's not appropriate (yet) to put into any development
> branch - and refer to that. That way we can do line-comments and such.
>
> I'm kind of surprised your msg.MarshalBinary() isn't creating an
> infinite-stack-depth loop: it passes a pointer to msg to abstract.Write,
> which might notice that msg implements the BinaryMarshaler interface can
> call MarshalBinary() on it instead of the "fallback" method of
> introspection, which will re-invoke your msg.MarshalBinary method, which
> calls abstract.Write on it again... I guess perhaps currently only the
> protobuf encoding mechanism consistently detects when objects to be written
> implement the BinaryMarshaler interface, but that's probably a bug in
> abstract.Write (and abstract.Read), which - when fixed - will make your
> code loop infinitely.
>
> In general, from this struct definition it doesn't like you should need to
> implement MarshalBinary() on the msg type at all; just let the caller
> (whoever wants the struct marshaled) pass the msg-pointer directly to
> abstract.Write, which will use reflection to figure out how to marshal it.
> You should only need to implement MarshalBinary() on types that *need* to
> "marshal themselves" for some reason and for which reflection-based
> marshaling won't work (e.g., structs whose members are private and/or
> special in other ways, custom encodings, etc.).
>
> It's hard to actually help you debug the real marshaling/unmarshaling
> problem without being able to checkout the code myself and run it - can you
> commit it to a branch somewhere (perhaps ephemeral) with a test-function
> that produces the behavior you're seeing, so that I can run and play with
> it? Thanks.
>
> —
> Reply to this email directly or view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_DeDiS_crypto_issues_35-23issuecomment-2D75137310&d=AwMCaQ&c=-dg2m7zWuuDZ0MUcV7Sdqw&r=GPxwT4fhhmVz-1RB6SMPpOGPGmZuqB1ZodgT3dSLWWI&m=__D0iYVL-PD_7Ixpn-AsAS562gHZqbShuaKYYN5eeww&s=LFZjv9z0jKyqsKWWQT2TP1u6dxYKHzQucaiFTo7p094&e=>
> .
>



-- 
William Edward Bailey, III

**Comment by bford** [comment]
OK, I just did a sanity-check by adding a println(len(secbytes)), and it prints 32, which is the size of one point or secret but makes no sense as the size of the whole polynomial.  Digging further.

**Comment by bford** [comment]
I need to go to a meeting, but what you need to do is dump the whole marshaled blob, sanity-check the length and visual look of it (hex.EncodeToString or hex.Dump is your friend), then add temporary printfs within the parts of the marshaling/unmarshaling code to trace through each step of the marshaling/unmarshaling process and make sure the right thing is happening and that the unmarshmaling is staying in sync with the marshaling, etc.  I suspect if you do this you'll figure out what's going on right away.

**Comment by WEB3-GForce** [comment]
I was able to do a similar thing with protobufs without an infinite loop. My PolicyApprovedMessage in policyMessage.go (insurance commit of prifi) uses protobufs and is successfully able to marshal / unmarshal. 

Concerning the infinite loop of the marshalling functions, would it still be good to provide the equivalent of the marshal/unmarshal functions for the user? Before the structure can be unmarshalled properly, the secret share and public polynomial would need to be initialized. The caller would have to do this every time that they wanted to unmarshal.

On that note, would it be useful to allow the messages to define unmarshal/ marshal and yet have abstract.Write also use introspection in select cases? We could signal to abstract.Write when we want to use the struct's marshal methods or use introspection. As mentioned earlier, it would be helpful to still be able to define marshal/ unmarshal methods that could do some set-up and then rely on introspection for the rest.

@dyv @iuliatamas,  I think it might be helpful for you all to look over the previous posts of Professor Ford since coco's signingMessages may run into such problems as the policyMessages.

**Comment by WEB3-GForce** [comment]
Sorry if my code wasn't clear: secbytes is the marshalled version of an abstract.Secret.  bytes is the marshalled version of the public polynomial. Hence, the printout is what is expected.

I'll debug the code further and see what I find.

**Comment by jackowitzd2** [comment]
Possibly relevant, ```abstract.Read/Write``` currently only special case the ```Marshaling``` interface, which you message does not implement - can you confirm that your marshaling functions are even getting called. You might be instead falling through to default encoding.

**Comment by bford** [comment]
Any progress on this?  Did you figure out what's going on?

**Comment by WEB3-GForce** [comment]
Working on it, but the bug is rather elusive. I've made some progress, but I don't know how to explain the results I've found.

To isolate the issue, I decided to create some test cases in sharing_test to focus specifically on the public polynomial. I've made an experimental branch and opened up a pull request if you wanted to see the code/ copy it.

While testing, I got the same error. The public polynomial is unable to validate shares after being decoded. Hence, the error seems to lie with the public polynomial encoding/ decoding.

The odd part is that the code reports the encoded and decoded versions as being the same. Their hex dump also produces the same results. When I check the abstract.Point's each contain, they are also equal. However, the check function fails.

Digging deeper, the issue is that the decoded polynomial produces the wrong result when evaluating the polynomial at point i. So, I reproduced the evaluation code comparing both the decoded and original versions at each step. The finding was rather surprising.

Everything is going fine until we get to the second element. Then, we diverge. I noticed that this was the same regardless of the k I tried. I can't explain the divergence. The points of the two polynomials used in the calculations are still equal. I'm not sure why the Mul and Add functions produce different results.

However, I came upon a chance find. So, I noticed that Eval starts from the end of the polynomial and evaluates backwards (j = k-1 down to j=0). I was originally checking whether the individual points of the polynomial were equal in the opposite direction (starting at i = 0) within the main loop of the test. So, I decided to check through all of them at the very beginning. Everything passed.

Wondering exactly why this happened, I noticed that what seems to be making everything work is this:

   if !testPubPoly.p[z].Equal(decodePubPoly.p[z]) ||
 		    testPubPoly.p[z].String() != decodePubPoly.p[z].String() {

This is where I compare the points to see if they are equal. If I comment out the string comparison code, it fails. If I keep it, everything passes.

I'm not sure why calling String fixes this or why the evaluations are diverging. 

I'll continue digging...
